queue.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2024 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package fifo
+
+import "sync"
+
+// Queue implements an allocation efficient FIFO queue. It is not safe for
+// concurrent access.
+//
+// -- Implementation --
+//
+// The queue is implemented as a linked list of nodes, where each node is a
+// small ring buffer. The nodes are allocated using a sync.Pool. It is
+// recommended to create a single pool for any given type and use it with all
+// queues of that type.
+type Queue[T any] struct {
```

Comment:
Is there a need for `Queue` to be exported? No strong objection to doing so, but I typically try to minimize the exposed API surface area.

Reply:
I think it's useful on its own. Maybe we'll make quotapool use it instead of notifyQueue

Reply:
It'd be better to extract the quotapool package than to have it adopt this thing, but it's fine either way. It is somewhat bespoke as a queue in that it provides access to the internal queue entry via the pointer returned in `PushBack`. Maybe it's worth noting something about that?